jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -2579,6 +2581,189 @@ ecma_builtin_array_prototype_object_copy_within (const ecma_value_t args[], /**<
+
+        targetIndex = ecma_builtin_array_flatten_into_array (target,
+                                                             element_obj,
+                                                             elementLen,
+                                                             targetIndex,
+                                                             depth-1,
+                                                             ECMA_VALUE_UNDEFINED,
+                                                             ECMA_VALUE_UNDEFINED);
+        ecma_deref_object (element_obj);
+      }
+      else
+      {
+        ecma_value_t element_temp = ecma_builtin_helper_def_prop_by_index (ecma_get_object_from_value (target),
+                                                                           targetIndex,
+                                                                           element,
+                                                                      ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE);
+
+        ecma_free_value (element);
+        if (ECMA_IS_VALUE_ERROR (element_temp))
+        {
+          return element_temp;
+        }
+
+        targetIndex++;
+      }
+    }
+    sourceIndex++;
+  }
+  return targetIndex;
+} /* ecma_builtin_array_flatten_into_array */
+
+/**
+ * The Array.prototype object's 'flat' routine
+ *
+ * See also:
+ *          ECMA-262 v10, 22.1.3.10
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_array_prototype_object_flat (const ecma_value_t args[], /**< arguments list */
+                                                ecma_length_t args_number, /**< number of arguments */
+                                                ecma_object_t *obj_p, /**< array object */
+                                                uint32_t len) /**< array object's length */
+{
+  ecma_number_t depthNum = 1;
+  if (args_number > 0)
+  {
+    if (ECMA_IS_VALUE_ERROR (ecma_op_to_integer (args[0], &depthNum)))
+    {
+      return ECMA_VALUE_ERROR;
+    }
+  }
+  ecma_value_t new_array = ecma_op_array_species_create (obj_p, 0);
+
+  if (ECMA_IS_VALUE_ERROR (new_array))
+  {
+    return new_array;
+  }
+  ecma_builtin_array_flatten_into_array (new_array,
+                                         obj_p,
+                                         len,
+                                         0,
+                                         depthNum,
+                                         ECMA_VALUE_UNDEFINED,
+                                         ECMA_VALUE_UNDEFINED);
+
+  return new_array;
+} /* ecma_builtin_array_prototype_object_flat */
+
+/**
+ * The Array.prototype object's 'flatMap' routine
+ *
+ * See also:
+ *          ECMA-262 v10, 22.1.3.11
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_array_prototype_object_flat_map (ecma_value_t arg1, /**< callbackFn */
```

Comment:
`argX` is not really descriptive. `callback` and `this_arg` are prefectly fine for variable names